DKP-312: Add release notes for 4.12.0 (#15503)
docker/docker.github.io: desktop/release-notes.md

For frequently asked questions about Docker Desktop releases, see [FAQs](faqs/general.md/#releases)

## Docker Desktop 4.12.0
2022-09-01

> Download Docker Desktop
>
> {% include desktop-install.html %}

## New

- Added the experimental docker daemon with the containerd image store.
- Allow Docker Desktop to run untagged images. Fixes [https://github.com/docker/for-mac/issues/6425]
- Add search capabilities to Docker Extension Marketplace. Fixes [docker/roadmap#346](https://github.com/docker/roadmap/issues/346).

> - Copied terminal contents do not contain non-breaking spaces anymore.

Author: Will move it

> -  Compose V2 is now enabled by default on new installations of Docker Desktop.

Reviewer: ```suggestion
- Compose V2 is now enabled by default on new installations of Docker Desktop.
- Precedence order of environment variables in Compose is more consistent, and clearly [documented](https://docs.docker.com/compose/envvars-precedence/).
```

Sorry for the late addition to the release notes! The updated version of Compose included in this release will include some changes to the implementation of environment variables. 

In the transition from v1 to v2, we received some user feedback that existing env var usage wasn't resulting in expected behavior. Since the evolution of the env vars implementation in v1 was quite organic, the rewrite in Go for v2 resulted in subtle changes. Instead of replicating the exact edge cases from v1, we address this by clearly documenting the precedence order in the latest version of Compose. 

It's debatable whether this is actually "New," but I placed the item here for now since we want users to be aware of this change. If you think it's more appropriate for another section, feel free to move it.

Reviewer 2: Thanks for the addition Han. Wording LGTM but think this should be moved to the 'Bugs and minor changes' section @andrei-z-albu

> - Enabled Compose V2 by default on new installations of Docker Desktop.

Reviewer 2: ```suggestion
-  Compose V2 is now enabled by default on new installations of Docker Desktop.
```

> - Added View menu with Zoom In, Zoom Out, and Actual Size options to control the size of the Docker Desktop UI.

Reviewer 2: Maybe i missed something, but i can't see this feature in DD 🤔

Author: On mac there's a new View menu next to Docker Desktop and Edit at the top, with the available actions mentioned here. I assume this shows up on win and linux in a similar fashion: ![Screenshot 2022-08-31 at 23 08 02](https://user-images.githubusercontent.com/54431776/187773486-bfbaee1a-8520-4232-ad83-050d45a85269.png)

Reviewer 2: Ah ok, I have Windows so can't see this. Think this should be moved to the Mac section in 'Bugs and minor changes'

Reviewer 3: The Zoom In, Zoom Out, and Actual Size also works for Windows with the keyboard shortcuts CTRL +/CTRL -/CTRL 0. So the feature itself works for all, but Windows Docker Dashboard doesn't have any menu like on Mac.
Maybe we should rephrase it a bit, because it's still an useful feature on Windows as well.

Reviewer 2: How about:

Added the ability to zoom in, out or set Docker Desktop to Actual Size. This is through the **View** menu on Mac or using keyboard shortcuts `CTRL +`, `CTRL -`, `CTRL 0` on Windows

Author: Minor suggestion to include shortcuts on mac as well: Added the ability to zoom in, out or set Docker Desktop to Actual Size. This is done by using keyboard shortcuts `⌘ +` / `CTRL +`, `⌘ -` / `CTRL -`, `⌘ 0` / `CTRL 0` on Mac and Windows respectively, or through the View menu on Mac.

Reviewer 2: Perfect 👍

- Added compose stop button if any related container is stoppable.

> - Individual compose containers are now deletable from overview.

Reviewer 2: ```suggestion
- Individual compose containers are now deletable from the **Container** view. 
```

> - Remove workaround for virtiofsd <-> qemu protocol mismatch on Fedora 35, as it is no longer needed. Fedora 35 users should upgrade the qemu package to the most recent version (qemu-6.1.0-15.fc35 as of the time of writing).

Reviewer 2: ```suggestion
- Removed the workaround for virtiofsd <-> qemu protocol mismatch on Fedora 35, as it is no longer needed. Fedora 35 users should upgrade the qemu package to the most recent version (qemu-6.1.0-15.fc35 as of the time of writing).
```

- Implemented an integrated terminal for containers.

> - Show address for external links.

Reviewer 2: ```suggestion
- Added a tooltip that displays the link address for all external links by default.
```

### Upgrades

- [Docker Compose v2.10.2](https://github.com/docker/compose/releases/tag/v2.10.2)
- [Docker Compose "Cloud Integrations" v1.0.29](https://github.com/docker/compose-cli/releases/tag/v1.0.29)
- [Docker Scan v0.19.0](https://github.com/docker/scan-cli-plugin/releases/tag/v0.19.0)
- [Kubernetes v1.25.0](https://github.com/kubernetes/kubernetes/releases/tag/v1.25.0)
- [Go 1.19](https://github.com/golang/go/releases/tag/go1.19)
- [cri-dockerd v0.2.5](https://github.com/Mirantis/cri-dockerd/releases/tag/v0.2.5)
- [Buildx v0.9.1](https://github.com/docker/buildx/releases/tag/v0.9.1)
- [containerd v1.6.8](https://github.com/containerd/containerd/releases/tag/v1.6.8)
- [containerd v1.6.7](https://github.com/containerd/containerd/releases/tag/v1.6.7)
- [runc v1.1.4](https://github.com/opencontainers/runc/releases/tag/v1.1.4)
- [runc v1.1.3](https://github.com/opencontainers/runc/releases/tag/v1.1.3)

### Bug fixes and minor changes

#### For all platforms

> - Enable Compose V2 after factory reset.

Reviewer 2: ```suggestion
- Compose V2 is now enabled after factory reset.
```

> - Upgraded kernel to 5.10.124

Reviewer 2: ```suggestion
- Upgraded kernel to 5.10.124.
```

- Improved overall performance issues caused by calculating disk size. Related to [docker/for-win#9401](https://github.com/docker/for-win/issues/9401).

> - Prevent users on ARM macs without Rosetta installed from switching back to Compose V1, which has only intel binaries.

Reviewer 2: ```suggestion
- Docker Desktop now prevents users on ARM macs without Rosetta installed from switching back to Compose V1, which has only intel binaries.
```

> - Descending default sort order for volumes size and created column, containers started column.

Reviewer 2: ```suggestion
- Changed the default sort order to descending for volume size and the **Created** column, along with the container's **Started** column.
```

- Re-organized container row actions.

> - Run every command on the Quickstart guide immediately.

Reviewer 2: ```suggestion
- The Quickstart guide now runs every command immediately.
```

> - Defined sort order for container/compose state column running > some running > paused > some paused > exited > some exited > created.

Reviewer 2: ```suggestion
- Defined the sort order for container/compose **Status** column to running > some running > paused > some paused > exited > some exited > created.
```

> - Fixed issues with the image list appearing empty in Docker Desktop even though there are images. Related to [docker/for-win#12693](https://github.com/docker/for-win/issues/12693) and [docker/for-mac#6347](https://github.com/docker/for-mac/issues/6347).

Reviewer 4: with this bug fix, i didn't note another change in behavior that could be noticeable to users. would it be possible to add it?
```
Defined what images are "in use" based on whether or not system containers are displayed; if system containers related to Kubernetes and Extensions are not displayed, the related images will not be defined as "in use."
```
(it's a difficult change to articulate succinctly, so feel free to modify if my phrasing is confusing!)

Reviewer 2: Small tweak to your suggestion:
Defined what images are "in use" based on whether or not system containers are displayed. If system containers related to Kubernetes and Extensions are not displayed, the related images are not defined as "in use."

Author: Added your revised description @aevesdocker

Reviewer 4: thank you both!

- Fixed a bug that made Docker clients in some languages hang on `docker exec`. Fixes [https://github.com/apocas/dockerode/issues/534](https://github.com/apocas/dockerode/issues/534).

> - Extensions SDK: prevent failed spawned command to make Docker Desktop unexpectedly quit.

Reviewer 2: ```suggestion
- A failed spawned command when building an extension no longer causes Docker Desktop to unexpectedly quit.
```

> - Fixed extensions being displayed as disabled in the left menu when they are not.

Reviewer 2: ```suggestion
- Fixed a bug that caused extensions to be displayed as disabled in the left menu when they are not.
```

- Fixed `docker login` to private registries when Registry Access Management is enabled and access to Docker Hub is blocked.

> - Fixed bug where we fail to start the Kubernetes cluster if the current cluster metadata is not stored in the `.kube/config` file.

Reviewer 2: ```suggestion
- Fixed a bug where Docker Desktop fails to start the Kubernetes cluster if the current cluster metadata is not stored in the `.kube/config` file.
```

- Updated the tooltip design.

#### For Mac

- Minimum version to install or update Docker Desktop on macOS is now 10.15. Fixes [docker/for-mac#6007](https://github.com/docker/for-mac/issues/6007).

> - Fixed a bug where Tray menu could incorrectly displays "Download will start soon..." after downloading the update. Fixes some issue reported in [for-mac/issues#5677](https://github.com/docker/for-mac/issues/5677)

Reviewer 2: ```suggestion
- Fixed a bug where the Tray menu incorrectly displays "Download will start soon..." after downloading the update. Fixes some issue reported in [for-mac/issues#5677](https://github.com/docker/for-mac/issues/5677)
```

> - UUpdated the tooltips in Docker Desktop and MUI theme package to align with the overall system design.

Reviewer 2: ```suggestion
- Updated the tooltips in Docker Desktop and MUI theme package to align with the overall system design.
```

- Fixed a bug that didn't restart Docker Desktop after applying an update.

> - When using virtualization.framework, fix bugs where connection to Docker is lost when the computer sleeps and when using restrictive firewall software.

Reviewer 2: ```suggestion
- Fixed a bug that caused the connection to Docker to be lost when the computer sleeps if a user is using virtualization.framework and restrictive firewall software.
```

> - Quit Docker Desktop completely from the Dock menu, to ensure Docker Desktop isn't running in the background without the whale icon in the menu bar visible. Fixes [https://github.com/docker/for-mac/issues/6440]

Reviewer 2: ```suggestion
- Fixed a bug that caused Docker Desktop to run in the background even after a user had quit the application.  Fixes [https://github.com/docker/for-mac/issues/6440]
```

- Disabled both Virtualization Framework and VirtioFS for users running macOS < 12.5

#### For Windows

- Fixed a bug where versions displayed during an update could be incorrect. Fixes [for-win/issues#12822](https://github.com/docker/for-win/issues/12822).

### Security 

#### For Windows

> - Fix a bypass for the `--no-windows-containers` installation flag which was introduced in version 4.11. This flag allows administrators to disable the use of Windows containers.

Reviewer 2: ```suggestion
- Fixed a bypass for the `--no-windows-containers` installation flag which was introduced in version 4.11. This flag allows administrators to disable the use of Windows containers.
```

> - Fix argument injection to the Docker Desktop installer which may result in local privilege escalation.

Reviewer 2: ```suggestion
- Fixed the argument injection to the Docker Desktop installer which may result in local privilege escalation.
```

> #### For all platforms

Reviewer 2: Can we put this section above the "For Windows" section?

For frequently asked questions about Docker Desktop releases, see [FAQs](faqs/general.md/#releases)

## Docker Desktop 4.12.0
2022-09-01

> Download Docker Desktop
>
> {% include desktop-install.html %}

### New

- Added the ability to use containerd for pulling and storing images. This is an experimental feature. 

> - Docker Desktop now runs untagged images. Fixes [https://github.com/docker/for-mac/issues/6425]

Reviewer 2: ```suggestion
- Docker Desktop now runs untagged images. Fixes [docker/for-mac#6425](https://github.com/docker/for-mac/issues/6425).
```

> [Mac with Apple chip](https://desktop.docker.com/mac/main/arm64/84025/Docker.dmg) |
> [Debian](https://desktop.docker.com/linux/main/amd64/84025/docker-desktop-4.11.0-amd64.deb) |
> [RPM](https://desktop.docker.com/linux/main/amd64/84025/docker-desktop-4.11.0-x86_64.rpm) |
> [Arch package](https://desktop.docker.com/linux/main/amd64/84025/docker-desktop-4.11.0-x86_64.pkg.tar.zst)

<div class="panel-group" id="accordion" role="tablist" aria-multiselectable="true">
  <div class="panel panel-default">
    <div class="panel-heading" role="tab" id="headingSeven">
      <h5 class="panel-title">
        <a role="button" data-toggle="collapse" data-parent="#accordion" href="#collapseSeven" aria-expanded="true" aria-controls="collapseSeven">
          Checksums
          <i class="fa fa-chevron-down"></i>
        </a>
      </h5>
    </div>
    <div id="collapseSeven" class="panel-collapse collapse" role="tabpanel" aria-labelledby="headingSeven">
      <div class="panel-body">
      <li><b>Windows:</b> SHA-256 8af32948447ddab655455542f6a12c8d752642a2bd451e2a48f76398cfd872b0</li>
      <li><b>Mac Intel:</b> SHA-256 b2f4ad8fea37dfb7d9147f169a9ceab71d7d0d12ff912057c60b58c0e91aed35</li>

>       <li><b>Mac Arm:</b> SHA-256 a7d84117bef83764cb9bf275cd01b8ba0c43f08dbfe4d4a7d4f05549cdd81f54</li>

Reviewer 3: Are the checksums up to date? I only verified this one on build 85629 and I got a different SHA-256 sum. 🤔

For frequently asked questions about Docker Desktop releases, see [FAQs](faqs/general.md/#releases)

## Docker Desktop 4.12.0
2022-09-01

> Download Docker Desktop
>
> {% include desktop-install.html %}

> ## New

Reviewer 2: I think some of these can be redistributed to the Bug fixes and minor changes sections. I'd suggest moving:

- Minimum version to install or update Docker Desktop on macOS is now 10.15
-Improved overall performance issues caused by calculating disk size. Related to [docker/for-win#9401](https://github.com/docker/for-win/issues/9401).
-  Enable ComposeV2 after factory reset
- Upgrade kernel to 5.10.124
- Prevents users on ARM macs without Rosetta installed from switching back to ComposeV1 (only has intel binaries).
- - Descending default sort order for volumes size and created column, containers started column
- Defined sort order for container/compose state column running > some running > paused > some paused > exited > some exited > created
- Re-organized container row actions
- - Runs every command on the Quickstart guide immediately.
- Updated the tooltip design.

Author: RIght, thanks! I have moved these

> - Minimum version to install or update Docker Desktop on macOS is now 10.15

Reviewer 5: I have just realized that there was a for-mac issue related to this:
```suggestion
  - Minimum version to install Docker Desktop on macOS is now 10.15. Fixes [docker/for-mac#6007](https://github.com/docker/for-mac/issues/6007)
```

- Improved overall performance issues caused by calculating disk size. Related to [docker/for-win#9401](https://github.com/docker/for-win/issues/9401).
- Added the experimental docker daemon with the containerd image store.

> - Allow Docker Desktop to run untagged image (https://github.com/docker/for-mac/issues/6425)

Reviewer 5: ```suggestion
- Allow Docker Desktop to run untagged images (https://github.com/docker/for-mac/issues/6425)
```
?

- Add search capabilities to Docker Extension Marketplace. Fixes [docker/roadmap#346](https://github.com/docker/roadmap/issues/346).

> - [mac] Quit Docker Desktop completely from the Dock menu (https://github.com/docker/for-mac/issues/6440)

Reviewer 5: That looks like a bug fix for the Mac section below?
Also the phrasing is a little odd…

Author: You're right, I'll move it to the Mac section. Not sure about the phrasing though, perhaps make it much more explicit like:Quit Docker Desktop completely from the Dock menu, to ensure Docker Desktop isn't running in the background without the whale icon in the menu bar visible. Fixes [https://github.com/docker/for-mac/issues/6440]

- Enable ComposeV2 after factory reset
- Copied terminal contents do not contain non-breaking spaces anymore

> - Upgrade kernel to 5.10.124

Reviewer 5: I believe all entries should be in the past form:
```suggestion
- Upgraded kernel to 5.10.124
```
However this one should be in the `Upgrades` section as
```
[Linux kernel 5.10.104](https://hub.docker.com/layers/for-desktop-kernel/docker/for-desktop-kernel/5.10.124-78a158e1ce5b203564f7436fda87b288e257d506/images/sha256-2faf709b9fe0962e4a151df6b4b48d2cf5636619ebf970584c6559ef0c1df940?context=explore)
```

> - Enables ComposeV2 by default on new installations/if there is no previous configuration.

Reviewer 5: ```suggestion
- Enabled Compose V2 by default on new installations of Docker Desktop.
```

- Prevents users on ARM macs without Rosetta installed from switching back to ComposeV1 (only has intel binaries).
- Added View menu with Zoom In, Zoom Out, and Actual Size options to control the size of the Docker Desktop UI.
- Descending default sort order for volumes size and created column, containers started column
- Defined sort order for container/compose state column running > some running > paused > some paused > exited > some exited > created

> - Re-organized container row actions

Reviewer 5: I think that change doesn’t mean much for users, unless we can describe more in details what this is about?

Author: Hmm, looks to me like most actions have been moved to the expand menu in the Action row, other than start/stop and delete. I don't know what 'pin action column' means from the [PR](https://github.com/docker/pinata/pull/18543). I have installed the new RC and this column doesn't seem pinned to me the way I imagine pinning should work. Do you think we should briefly describe the moved action items though?
Something like: Re-organized container row actions by keeping only the start/stop and delete actions visible at all times, while allowing access to the rest via the row menu item.

Reviewer 2: Your suggestion LGTM @andrei-z-albu 👍

- Added compose stop button if any related container is stoppable.
- Individual compose containers are now deletable from overview
- Runs every command on the Quickstart guide immediately.
- Remove workaround for virtiofsd <-> qemu protocol mismatch on Fedora 35, as it is no longer needed. Fedora 35 users should upgrade the qemu package to the most recent version (qemu-6.1.0-15.fc35 as of the time of writing).
- Implemented an integrated terminal for containers.

> - Updated the tooltip design.

Reviewer 5: This should either be removed or we should explain a bit more what it means?

Author: Hmm, from what I can understand out of the PR, this styling change is done so we better adhere to our own UI guidelines. Perhaps a message such as this would be more relevant to users? (otherwise I agree we should just remove this entry):
Updates the tooltips in Docker Desktop and MUI theme package to align with the overall system design.

Reviewer 2: Like your suggestion @andrei-z-albu, would just change 'Updates' to 'Updated'

Author: Great! I'll change it thus

> - Show address for external links.

Reviewer 5: I have no idea what this means. 😅

Author: Taken from the [PR](https://github.com/docker/pinata/pull/18547), the description there sounds like a better candidate for our release notes: "Adds a tooltip that displays the link address for all external links by default". What do you think?

### Upgrades

- [Docker Compose v2.10.2](https://github.com/docker/compose/releases/tag/v2.10.2)

> - [Docker Compose "Cloud Integrations" v1.0.29](https://github.com/docker/compose-cli/releases/tag/v1.0.29)

Reviewer 5: I don’t think we’ve been listing that component in the release notes in the past, have we?

Reviewer 2: If it's not in previous release note patterns, I'd remove it

- [Docker Scan v0.19.0](https://github.com/docker/scan-cli-plugin/releases/tag/v0.19.0)
- [Kubernetes v1.25.0](https://github.com/kubernetes/kubernetes/releases/tag/v1.25.0)
- [Go 1.19](https://github.com/golang/go/releases/tag/go1.19)
- [cri-dockerd v0.2.5](https://github.com/Mirantis/cri-dockerd/releases/tag/v0.2.5)
- [Buildx v0.9.1](https://github.com/docker/buildx/releases/tag/v0.9.1)

> - [containerd v1.6.8](https://github.com/containerd/containerd/releases/tag/v1.6.8)
> - [containerd v1.6.7](https://github.com/containerd/containerd/releases/tag/v1.6.7)

Reviewer 5: Only one is probably enough?
```suggestion
- [containerd v1.6.8](https://github.com/containerd/containerd/releases/tag/v1.6.8)
```

Author: Feels to me that one bullet with two links not would be as readable as two separate bullets, assuming we want to showcase both versions in the release notes. But I don't have a strong opinion about it

Author: What do you think @thaJeztah and @mat007 , shall we leave both entries in?

> - [runc v1.1.4](https://github.com/opencontainers/runc/releases/tag/v1.1.4)
> - [runc v1.1.3](https://github.com/opencontainers/runc/releases/tag/v1.1.3)

Reviewer 5: Same here
```suggestion
- [runc v1.1.4](https://github.com/opencontainers/runc/releases/tag/v1.1.4)
```

Reviewer 6: See above

### Bug fixes and minor changes

#### For all platforms

- Fixed issues with the image list appearing empty in Docker Desktop even though there are images. Related to [docker/for-win#12693](https://github.com/docker/for-win/issues/12693) and [docker/for-mac#6347](https://github.com/docker/for-mac/issues/6347).
- Fixed a bug that made Docker clients in some languages hang on `docker exec`. Fixes [https://github.com/apocas/dockerode/issues/534](https://github.com/apocas/dockerode/issues/534).
- Extensions SDK: prevent failed spawned command to make Docker Desktop unexpectedly quit.

> - Fixes container terminal resizing issues.

Reviewer 5: That’s odd because further up we have an entry saying we have just added that as a new feature.
We don’t need to document bug fixes for not yet released features. 😄

Author: Oh, right, that is odd to leave in there. Will take it out :D

> - Fixes extensions being displayed as disabled in the left menu but they are not.

Reviewer 5: ```suggestion
- Fixed extensions being displayed as disabled in the left menu when they are not.
```

> - Fix `docker login` to private registries when Registry Access Management is enabled and access to Docker Hub is blocked.

Reviewer 5: ```suggestion
- Fixed `docker login` to private registries when Registry Access Management is enabled and access to Docker Hub is blocked.
```

- Fix bug where we fail to start the Kubernetes cluster if the current cluster metadata is not stored in the `.kube/config` file.

#### For Mac

- Fixed a bug where Tray menu could incorrectly displays "Download will start soon..." after downloading the update. Fixes some issue reported in [for-mac/issues#5677](https://github.com/docker/for-mac/issues/5677)
- Fixed a bug that didn't restart Docker Desktop after applying an update.
- When using virtualization.framework, fix bugs where connection to Docker is lost when the computer sleeps and when using restrictive firewall software.

#### For Windows

- Fixed a bug where versions displayed during an update could be incorrect. Fixes [for-win/issues#12822](https://github.com/docker/for-win/issues/12822).

## Docker Desktop 4.11.1

> 2022-08-05

Reviewer 6: We need to remove the `{% include desktop-install.html %}` for 4.11.1 and replace it with a block with regular links (and checksum section) similar as the 4.11.0 release, otherwise it shows the "install" block twice for the latest release.

https://github.com/docker/docker.github.io/blame/d282ce0a6360e70d88131d37f814ee63cb785d5d/desktop/release-notes.md#L49-L77

Author: Thanks! Fixed with a new push
